handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -0,0 +1,1951 @@
+                // https://github.com/apache/httpd/blob/2.4.16/modules/ssl/ssl_engine_kernel.c#L812
+                // http://h71000.www7.hp.com/doc/83final/ba554_90007/ch04s03.html
+                if (SSL.renegotiate(ssl) != 1 || SSL.doHandshake(ssl) != 1) {
+                    throw shutdownWithError("renegotiation failed");
+                }
+
+                SSL.setState(ssl, SSL.SSL_ST_ACCEPT);
+
+                lastAccessed = System.currentTimeMillis();
+
+                // fall-through
+            case NOT_STARTED:
+                handshakeState = HandshakeState.STARTED_EXPLICITLY;
+                handshake();
+                break;
+            default:
+                throw new Error();
+        }
+    }
+
+    private void checkEngineClosed(SSLException cause) throws SSLException {
+        if (engineClosed || isDestroyed()) {
+            throw cause;
+        }
+    }
+
+    private static SSLEngineResult.HandshakeStatus pendingStatus(int pendingStatus) {
+        // Depending on if there is something left in the BIO we need to WRAP or UNWRAP
+        return pendingStatus > 0 ? NEED_WRAP : NEED_UNWRAP;
+    }
+
+    private SSLEngineResult.HandshakeStatus handshake() throws SSLException {
+        if (handshakeState == HandshakeState.FINISHED) {
+            return FINISHED;
+        }
+        checkEngineClosed(HANDSHAKE_ENGINE_CLOSED);
+
+        // Check if we have a pending handshakeException and if so see if we need to consume all pending data from the
+        // BIO first or can just shutdown and throw it now.
+        // This is needed so we ensure close_notify etc is correctly send to the remote peer.
+        // See https://github.com/netty/netty/issues/3900
+        SSLHandshakeException exception = handshakeException;
+        if (exception != null) {
+            if (SSL.pendingWrittenBytesInBIO(networkBIO) > 0) {
+                // There is something pending, we need to consume it first via a WRAP so we not loose anything.
+                return NEED_WRAP;
+            }
+            // No more data left to send to the remote peer, so null out the exception field, shutdown and throw
+            // the exception.
+            handshakeException = null;
+            shutdown();
+            throw exception;
+        }
+
+        // Adding the OpenSslEngine to the OpenSslEngineMap so it can be used in the AbstractCertificateVerifier.
+        engineMap.add(this);
+        if (lastAccessed == -1) {
+            lastAccessed = System.currentTimeMillis();
+        }
+
+        if (!certificateSet && keyMaterialManager != null) {
+            certificateSet = true;
+            keyMaterialManager.setKeyMaterial(this);
+        }
+
+        int code = SSL.doHandshake(ssl);
+        if (code <= 0) {
+            // Check if we have a pending exception that was created during the handshake and if so throw it after
+            // shutdown the connection.
+            if (handshakeException != null) {
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Change this condition so that it does not always evaluate to "false" [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2583)